internal/ethapi/tracer.go

```diff
@@ -0,0 +1,321 @@
+}
+
+// length returns the length of the stack
+func (sw *stackWrapper) length() int {
+	return len(sw.stack.Data())
+}
+
+// toValue returns an otto.Value for the stackWrapper
+func (sw *stackWrapper) toValue(vm *otto.Otto) otto.Value {
+	value, _ := vm.ToValue(sw)
+	obj := value.Object()
+	obj.Set("peek", sw.peek)
+	obj.Set("length", sw.length)
+	return value
+}
+
+// dbWrapper provides a JS wrapper around vm.Database
+type dbWrapper struct {
+	db vm.Database
+}
+
+// getBalance retrieves an account's balance
+func (dw *dbWrapper) getBalance(addr common.Address) *big.Int {
+	return dw.db.GetBalance(addr)
+}
+
+// getNonce retrieves an account's nonce
+func (dw *dbWrapper) getNonce(addr common.Address) uint64 {
+	return dw.db.GetNonce(addr)
+}
+
+// getCode retrieves an account's code
+func (dw *dbWrapper) getCode(addr common.Address) []byte {
+	return dw.db.GetCode(addr)
+}
+
+// getState retrieves an account's state data for the given hash
+func (dw *dbWrapper) getState(addr common.Address, hash common.Hash) common.Hash {
+	return dw.db.GetState(addr, hash)
+}
+
+// exists returns true iff the account exists
+func (dw *dbWrapper) exists(addr common.Address) bool {
+	return dw.db.Exist(addr)
+}
+
+// toValue returns an otto.Value for the dbWrapper
+func (dw *dbWrapper) toValue(vm *otto.Otto) otto.Value {
+	value, _ := vm.ToValue(dw)
+	obj := value.Object()
+	obj.Set("getBalance", dw.getBalance)
+	obj.Set("getNonce", dw.getNonce)
+	obj.Set("getCode", dw.getCode)
+	obj.Set("getState", dw.getState)
+	obj.Set("exists", dw.exists)
+	return value
+}
+
+// JavascriptTracer provides an implementation of Tracer that evaluates a
+// Javascript function for each VM execution step.
+type JavascriptTracer struct {
+	vm         *otto.Otto             // Javascript VM instance
+	traceobj   *otto.Object           // User-supplied object to call
+	log        map[string]interface{} // (Reusable) map for the `log` arg to `step`
+	logvalue   otto.Value             // JS view of `log`
+	opcode     *opCodeWrapper         // Wrapper around the opcode
+	opcodevalue otto.Value  		  // JS view of 'opcode'
+	memory     *memoryWrapper         // Wrapper around the VM memory
+	memvalue   otto.Value             // JS view of `memory`
+	stack      *stackWrapper          // Wrapper around the VM stack
+	stackvalue otto.Value             // JS view of `stack`
+	db         *dbWrapper             // Wrapper around the VM environment
+	dbvalue    otto.Value             // JS view of `db`
+	err        error                  // Error, if one has occurred
+	timeout    time.Duration          // Remaining execution time
+}
+
+// NewJavascriptTracer instantiates a new JavascriptTracer instance.
+// code specifies a Javascript snippet, which must evaluate to an expression
+// returning an object with 'step' and 'result' functions.
+func NewJavascriptTracer(code string, timeout time.Duration) (*JavascriptTracer, error) {
+	vm := otto.New()
+	vm.Interrupt = make(chan func(), 1)
```

Comment:
It would be better to use an overall deadline for the entire tracing operation. It would be best to use
a context that is passed down from the request. So `NewJavascriptTracer` would take a context and use ctx.Done() as the interrupt channel.

This approach frees you from taking care of the timer.